rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+at the root of the target. Given the following example:
+
+```json
+{ "foo": { "bar": true } }
+```
+
+How would we use dynamic paths in a recursive for-loop?
+
+```coffee
+recursive for (key, value) in . {
+  .[upcase(key)] = value
+}
+```
+
+Because key is `"foo"` and then `"bar"`, you would end up with:
+
+```json
+{ "FOO": true, "BAR": true }
+```
+
+Which is not the expected outcome.
+
+This could be solved by making `.` relative in the for-loop, but that's a major
+shift from the current way VRL works, requires a new way of accessing the root
+object if you can't use `.`, and goes against the rules as laid out in the
+[design document][doc].
+
+---
+
+- What other approaches have been considered and why did you not choose them?
+- How about not doing this at all?
+
+## Outstanding Questions
+
+- Do we want to introduce any form of lexical-scoping in this RFC, or keep the
+  status-quo for now?
+- Do we want to introduce tuple-expressions or are we satisfied with using
+  a two-element array as the return value for `map`?
+- ...
+
+## Plan Of Attack
+
+TODO
+
+## Future Improvements
+
+#### Iteration Control-Flow
+
+While likely desirable, this RFC intentionally avoids control-flow operations
+inside iterators.
+
+They are likely to be one of the first enhancements to this feature, though:
+
+```coffee
+. = map(.) |key, value| {
+  # Return default key/value pairs if the value is an object.
+  if is_object(value) {
+    return [key, value]
+  }
+
+  # ...
+}
+```
+
+#### Specialized Iteration Functions
+
+Once this RFC is implemented, additional iteration capability can be expanded by
+adding new functions to the standard library.
+
+For example, filtering:
+
+```coffee
+# Return a new array with "180.14.129.174" removed.
+.ips = filter(.ips) |_index, ip| {
+    ip = string(ip) ?? "unknown"
+
+    !starts_with(ip, "180.14")
+}
+```
+
+Or ensuring all elements adhere to a condition:
+
+```coffee
+# Add new `all_public` boolean field.
+.all_public = all(.ips) |_index, ip| {
+    ip = string(ip) ?? "unknown"
+
+    !starts_with(ip, "180.14")
+}
+```
+
+Some additional suggestions include `flatten`, `partition`, `fold`, `any`,
+`find`, `max`, `min`, etc...
+
+#### Schema Support
+
+Once [schema support][] lands, writing iterators can become less verbose.
+
+For example, this example from the RFC:
+
+```coffee
+.ips = array(.ips) ?? []
+.ips = filter(.ips) |_index, ip| {
+    ip = string(ip) ?? "unknown"
+
+    !starts_with(ip, "180.14")
+}
+```
+
+Can be written as follows, when applying the correct schema:
+
+```coffee
+.ips = filter(.ips) |_, ip| !starts_with(ip, "180.14")
+```
+
+Because a type schema could guarantee the compiler that `.ips` is an array, with
+only string items.
+
+#### Pipeline Operator Support
+
+Once the [pipeline operations][] land, we can further expand the above example
+as follows:
+
+```coffee
+.private_and_public_ips = filter(.ip) |_, ip| is_ip(ip) |> partition() |_, ip| starts_with(ip, "180.14")
```

Comment:
Regarding the discussion in https://github.com/timberio/vector/pull/9001/files#r700935702, wouldn't this case here be impossible to disambiguate when parsing?

Specifically, if `|>` relates to `is_ip(ip)` or `filter(.ip) |_, ip| is_ip(ip)`

Reply:
It wouldn't. It would require defining operator precedence, but we already do that.

That is, the pipeline operator would either be part of the inner expression, or it wouldn't. In the above example, it _wouldn't_.

Having said that, I agree it's at least ambiguous to readers, so that would be an argument against allowing this short-hand form of writing closures.

I think it's still nice to do (for example):

```coffee
. = map(.) |k, v| [upcase(k), v]
```

But it's not _terrible_ if that weren't possible:

```coffee
. = map(.) |k, v| { [upcase(k), v] }
```

Reply:
That's a good catch - does seem like we'll need some sort of identifier here `->` or `{}`

Reply:
Resolved by changing the function-closure syntax, which also removes the short-hand notation. See 1ad39083701c4c6739715490435d761e3a1eecb5.